From 271ac3927f0371db13ecb5adac46dfd0fb3aa687 Mon Sep 17 00:00:00 2001
From: Noam Camus <ncamus@habana.ai>
Date: Thu, 15 Nov 2018 15:06:47 +0200
Subject: [PATCH] SPI: update methods

We add flash protection preperation
Also we use more generic approach same as we use in hl-fw-loader

Change-Id: I19a870b666fa661b2d293386147d589b90ef30be
---

diff --git a/preboot.c b/preboot.c
index d076c2d..4dbfda2 100644
--- a/preboot.c
+++ b/preboot.c
@@ -10,6 +10,8 @@
 #include <linux/io.h>
 #include <linux/sizes.h>
 #include "mach/hl_regs.h"
+#include <stddef.h>
+#include <stdbool.h>
 
 #define SYS_REF_CLOCK			50000000
 #define BTL_COUNTER_MAX			BIT(22)
@@ -22,10 +24,26 @@
 #define mmPSOC_SPI_BAUDR		(HL_CFG_BASE + 0xC43014)
 #define mmPSOC_SPI_TXFTLR		(HL_CFG_BASE + 0xC43018)
 #define mmPSOC_SPI_RXFTLR		(HL_CFG_BASE + 0xC4301C)
+#define mmPSOC_SPI_TXFLR		(HL_CFG_BASE + 0xC43020)
 #define mmPSOC_SPI_RXFLR		(HL_CFG_BASE + 0xC43024)
 #define mmPSOC_SPI_IMR			(HL_CFG_BASE + 0xC4302C)
 #define mmPSOC_SPI_DR0			(HL_CFG_BASE + 0xC43060)
 #define SPL_SIZE			SZ_64K
+#define CMD_WRITE_STATUS		0x01
+#define CMD_READ			0x03
+#define CMD_WRITE_DISABLE		0x04
+#define CMD_READ_STATUS			0x05
+#define CMD_WRITE_ENABLE		0x06
+#define CMD_READ_CONF			0x15
+#define READ_STATUS_WIP			BIT(0)
+#define READ_STATUS_BP0			BIT(2)
+#define READ_STATUS_BP1			BIT(3)
+#define READ_STATUS_BP2			BIT(4)
+#define READ_STATUS_BP3			BIT(5)
+#define READ_STATUS_SRWD		BIT(7)
+#define READ_STATUS_BP0_3		(READ_STATUS_BP0 | READ_STATUS_BP1 | READ_STATUS_BP2 | READ_STATUS_BP3)
+#define PROT_MASK			(READ_STATUS_BP0_3 | READ_STATUS_SRWD)
+#define frame_size_bits(b)		((b) << 16)
 
 /* Use DumpSel[8] to skip PCIe initialization */
 #ifdef DEV_IS_GOYA
@@ -123,45 +141,153 @@
 	writel(0, mmPSOC_GLOBAL_CONF_PAD_SEL_53);
 }
 
-static void hl_spi_init(unsigned int offset)
+static inline void spi_dev_update_frame_size(unsigned int len)
 {
-	unsigned int cmd = GENMASK(25,24); /* Read command */
-	unsigned int bytes_per_frame = sizeof(unsigned int);
-	unsigned int frame_size_bit = (bytes_per_frame * 8) - 1;
-	unsigned int frame_size = frame_size_bit << 16;
+	writel(0, mmPSOC_SPI_SSIENR);
+	writel(frame_size_bits((len * 8) - 1), mmPSOC_SPI_CTRLR0);
+	writel(1, mmPSOC_SPI_SSIENR);
+}
 
+static inline int fifo_is_empty(unsigned long level_reg)
+{
+	return !readl(level_reg);
+}
+
+static void spi_dev_dword_cmd(unsigned int cmd, unsigned int *val)
+{
+	unsigned int tmp;
+
+	while (!fifo_is_empty(mmPSOC_SPI_TXFLR));
+	writel(cmd, mmPSOC_SPI_DR0);
+
+	while (fifo_is_empty(mmPSOC_SPI_RXFLR));
+	tmp = readl(mmPSOC_SPI_DR0);
+
+	if (val)
+		*val = tmp;
+}
+
+static void hl_spi_special_cmd(unsigned int cmd, unsigned char *val)
+{
+	unsigned char tmp;
+
+	spi_dev_update_frame_size(sizeof(tmp));
+
+	while (!fifo_is_empty(mmPSOC_SPI_TXFLR));
+	writeb(cmd, mmPSOC_SPI_DR0);
+
+	while (fifo_is_empty(mmPSOC_SPI_RXFLR));
+	tmp = readb(mmPSOC_SPI_DR0);
+
+	if (val)
+		*val = tmp;
+}
+
+/* Hold chip select (CS) during transaction */
+static void hl_spi_char_cmd_csel(unsigned char cmd, unsigned char *val)
+{
+	board_spi_claim_bus();
+
+	hl_spi_special_cmd(cmd, NULL);
+	if (val)
+		hl_spi_special_cmd(0, val);
+
+	board_spi_release_bus();
+}
+
+static void hl_spi_special_cmd_csel(unsigned int cmd, unsigned char *val, size_t size)
+{
+	int i;
+	unsigned char tmp;
+
+	board_spi_claim_bus();
+
+	spi_dev_update_frame_size(sizeof(*val));
+
+	while (!fifo_is_empty(mmPSOC_SPI_TXFLR));
+	writeb(cmd, mmPSOC_SPI_DR0);
+	for (i=0; i<size; i++) {
+		while (!fifo_is_empty(mmPSOC_SPI_TXFLR));
+		writeb(val[i], mmPSOC_SPI_DR0);
+	}
+
+	while (fifo_is_empty(mmPSOC_SPI_RXFLR));
+	tmp = readb(mmPSOC_SPI_DR0);
+	for (i=0; i<size; i++) {
+		while (fifo_is_empty(mmPSOC_SPI_RXFLR));
+		tmp = readb(mmPSOC_SPI_DR0);
+	}
+
+	board_spi_release_bus();
+}
+
+static void hl_spi_common_write(unsigned int cmd, unsigned char *buf, size_t size)
+{
+	unsigned char sr;
+
+	hl_spi_char_cmd_csel(CMD_WRITE_ENABLE, NULL);
+	hl_spi_special_cmd_csel(cmd, buf, size);
+	do
+		hl_spi_char_cmd_csel(CMD_READ_STATUS, &sr);
+	while (sr & READ_STATUS_WIP);
+	hl_spi_char_cmd_csel(CMD_WRITE_DISABLE, NULL);
+}
+
+static unsigned char hl_spi_prot(void)
+{
+	unsigned char sr, cr, data[2];
+
+	hl_spi_char_cmd_csel(CMD_READ_CONF, &cr);
+	hl_spi_char_cmd_csel(CMD_READ_STATUS, &sr);
+	sr |= PROT_MASK;
+	data[0] = sr;
+	data[1] = cr;
+	hl_spi_common_write(CMD_WRITE_STATUS, data, sizeof(data));
+	return sr;
+}
+
+static void hl_spi_read_spl(unsigned int offset)
+{
+	void *dst = (void*)CONFIG_SPL_TEXT_BASE;
+	unsigned int cmd = (CMD_READ << 24) | offset;
+	int i;
+
+	board_spi_claim_bus();
+	spi_dev_update_frame_size(sizeof(cmd));
+	spi_dev_dword_cmd(cmd, NULL);
+	for (i=0; i < SPL_SIZE; i += sizeof(unsigned int))
+		spi_dev_dword_cmd(0, dst + i);
+	spi_dev_dword_cmd(0, NULL);
+	board_spi_release_bus();
+}
+
+static void hl_spi_init(void)
+{
 	hl_ssi_cs0_init();
 	writel(0x00000000, mmPSOC_SPI_SSIENR);
-	writel(frame_size, mmPSOC_SPI_CTRLR0);
-	writel(0x00000002, mmPSOC_SPI_BAUDR);
+	writel(0x00000002, mmPSOC_SPI_BAUDR);	/* Min divisor */
 	writel(0x00000000, mmPSOC_SPI_RXFTLR);
 	writel(0x00000000, mmPSOC_SPI_TXFTLR);
-	writel(0x0000003F, mmPSOC_SPI_IMR); /* Disable Interrupts */
+	writel(0x0000003F, mmPSOC_SPI_IMR);	/* Disable Interrupts */
 	writel(0x00000001, mmPSOC_SPI_SER);
 	writel(0x00000001, mmPSOC_SPI_SSIENR);
-	writel(cmd | offset, mmPSOC_SPI_DR0);
-	while (!readl(mmPSOC_SPI_RXFLR));
-	cmd = readl(mmPSOC_SPI_DR0); /* Dummy read */
 }
 
 static void hl_preboot_copy_spl(void)
 {
-	void *spl_dst;
-	unsigned int i, spl_offset;
+	unsigned int  spl_offset;
+	unsigned char sr;
 
 	hl_preboot_copy_prepare(&spl_offset);
-	spl_dst = (void*)CONFIG_SPL_TEXT_BASE;
 
-	board_spi_claim_bus();
-	hl_spi_init(spl_offset);
-	for (i=0; i < SPL_SIZE; i += sizeof(unsigned int)) {
-		writel(0, mmPSOC_SPI_DR0); /* Dummy write */
-		while (!readl(mmPSOC_SPI_RXFLR));
-		*(unsigned int*)(spl_dst + i) = readl(mmPSOC_SPI_DR0);
-	}
-	board_spi_release_bus();
-
+	hl_spi_init();
+	hl_spi_read_spl(spl_offset);
+	sr = hl_spi_prot();
 	hl_preboot_puts("SPL copying from flash is completed\n");
+	if ((sr & PROT_MASK) == PROT_MASK)
+		hl_preboot_puts("Flash may be protected, depend on WP# signal\n");
+	else
+		hl_preboot_puts("Flash is not protected\n");
 }
 
 #ifdef DEV_IS_GOYA
